test/fw/ptl/utils/pbs_snaputils.py

```diff
@@ -0,0 +1,1777 @@
+        ATR_DFLAG_OPWR = 0x08
+        ATR_DFLAG_MGRD = 0x10
+        ATR_DFLAG_MGWR = 0x20
+        ATR_DFLAG_RASSN = 0x4000
+        ATR_DFLAG_ANASSN = 0x8000
+        ATR_DFLAG_FNASSN = 0x10000
+        ATR_DFLAG_CVTSLT = 0x20000
+
+        NO_USER_SET = (ATR_DFLAG_USRD | ATR_DFLAG_OPRD | ATR_DFLAG_MGRD |
+                       ATR_DFLAG_OPWR | ATR_DFLAG_MGWR)
+        READ_WRITE = (ATR_DFLAG_USRD | ATR_DFLAG_OPRD | ATR_DFLAG_MGRD |
+                      ATR_DFLAG_USWR | ATR_DFLAG_OPWR | ATR_DFLAG_MGWR)
+
+        resc_flag = READ_WRITE
+        if "q" in flag:
+            resc_flag |= ATR_DFLAG_RASSN
+        if "f" in flag:
+            resc_flag |= ATR_DFLAG_FNASSN
+        if "n" in flag:
+            resc_flag |= ATR_DFLAG_ANASSN
+        if "h" in flag:
+            resc_flag |= ATR_DFLAG_CVTSLT
+        if "r" in flag:
+            resc_flag &= ~READ_WRITE
+            resc_flag |= NO_USER_SET
+        if "i" in flag:
+            resc_flag &= ~READ_WRITE
+            resc_flag |= (ATR_DFLAG_OPRD | ATR_DFLAG_OPWR |
+                          ATR_DFLAG_MGRD | ATR_DFLAG_MGWR)
+        return resc_flag
+
+    def __convert_type_to_numeric(self, attr_type):
+        """
+        Convert a resource's type attribute to its numeric equivalent
+
+        :param attr_type: the type to convert
+        :type attr_type: string
+        :returns: Numeric equivalent of attr_type
+        """
+        PBS_ATTR_TYPE_TO_INT = {
+            "long": 1,
+            "string": 3,
+            "string_array": 4,
+            "size": 5,
+            "boolean": 11,
+            "float": 14,
+        }
+
+        return PBS_ATTR_TYPE_TO_INT[attr_type.strip()]
+
+    def __capture_trace_from_core(self, core_file_name, exec_path, out_path):
+        """
+        Capture stack strace from the core file specified
+
+        :param core_file_name: name of the core file
+        :type core_file_name: str
+        :param exec_path: path to the executable which generated the core
+        :type exec_path: str
+        :param out_path: ofile to print the trace out to
+        :type out_path: str
+        """
+        self.logger.debug("capturing stack trace from core file " +
+                          core_file_name)
+
+        # Create a gdb-python script to capture backtrace from core
+        gdb_python = "import gdb\n" \
+            "gdb.execute(\"core " + core_file_name + "\")\n" \
+            "o = gdb.execute(\"thread apply all bt\", to_string=True)\n" \
+            "print(o)\n" \
+            "gdb.execute(\"quit\")\n" \
+            "quit()"
+
+        # Open a temporary file to write the gdb-python script above
+        fd, fn = self.mkstemp(mode=0777)
+        with os.fdopen(fd, "w") as tempfd:
+            tempfd.write(gdb_python)
+
+        # Catch the stack trace using gdb
+        gdb_cmd = ["gdb", exec_path, "-P", fn]
+        with open(out_path, "w") as outfd:
+            Popen(gdb_cmd, stdout=outfd, stderr=STDOUT)
```

Comment:
You can make use of run_cmd here, because Popen is non-blocking and here we have to wait for gdb_cmd to finish.